[KB-201] Validation and Testing Procedures for Model Evaluation section (#157)
chhoumann/thesis-chemcam: report_thesis/src/sections/methodology/testing_validation.tex

However, each target is shot at multiple locations, resulting in multiple instances of the same target in the dataset, as shown in Table~\ref{tab:final_dataset_example}.
Although the intensity values vary for each location, they fundamentally represent different measurements of the same target.
If we were to randomly split the dataset, some locations from a target could end up in the testing set while others remain in the training set.
This would cause data leakage, as the testing set would no longer consist solely of unseen targets.
To prevent this, we ensure that each target is represented only once in the dataset by grouping data from all locations on a given target.

Furthermore, the limited availability we mention in Section~\ref{sec:problem_definition} of data poses another significant challenge.
The dataset we use consists of 408 samples. This is a large dataset by \gls{libs} standards.
As mentioned, the collection process is difficult, making datasets relatively small in general.
Additionally, there are only a few samples with extreme concentration values for the oxides in the targets.

> These extreme values act as outliers, presenting concentrations significantly different from the rest of the data points.

Author: > In statistics, an outlier is a data point that differs significantly from other observations. An outlier may be due to a variability in the measurement, an indication of novel data, or it may be the result of experimental error; the latter are sometimes excluded from the data set.

```suggestion
These extreme values act as benign outliers, presenting concentrations significantly different from the rest of the data points.
```

> These extreme values act as outliers, presenting concentrations significantly different from the rest of the data points.

Author: Resolved

When performing a random split of the dataset into multiple folds for cross-validation, as well as for training and testing sets, this small number of extreme values can result in an uneven distribution.
The presence or absence of these outliers in any given fold can heavily influence the model's performance metrics.
If extreme values are disproportionately allocated to either the training or testing sets, the resulting model may struggle to generalize accurately.
This uneven distribution can lead to models that perform well on the majority of the data but fail to predict accurately for these outlier concentrations, which are critical in many practical applications.

\begin{figure*}[h!]
    \centering
    \includegraphics[width=\textwidth]{images/oxide_distributions.png}

>     \caption{Distributions of various oxide concentrations in the dataset. The histograms show the frequency of concentration values for \ce{SiO2}, \ce{TiO2}, \ce{Al2O3}, \ce{FeOT}, \ce{MgO}, \ce{CaO}, \ce{Na2O}, and \ce{K2O}.}

Author: Nice catch 😵

Next, the dataset is processed to remove any duplicate entries based on the group column and then sorted by the target column (Line~\ref{line:remove_duplicates}).
This step ensures that each group is uniquely identified and ordered appropriately.
The dataset we illustrate in Table~\ref{tab:final_dataset_example} would require a group column $g$ of "\texttt{Target}" to group the data by target.
The target column $t$ refers to the column with the target variable, which would be the oxide for which we are predicting the concentration, e.g., \ce{SiO_2}.
By sorting the dataset by the target column, we ensure that the data is ordered by the target concentration values in ascending order.

Fold numbers are then assigned sequentially to the unique targets (Line~\ref{line:assign_folds}).
This initial assignment distributes the targets randomly across the folds.
If handling of extreme values is enabled, the algorithm identifies the top and bottom percentiles of the target values (Line~\ref{line:identify_extremes}) and reassigns these extreme values to the training folds (0 to \( k-2 \)) to ensure they are well-represented during model training (Line~\ref{line:reassign_extremes}).

> For our purposes, it is always enabled.

Reviewer: I dont necessarily think its bad to have the if statement, but if we decide to keep it, I think we should explain the case why you perhaps wouldnt want to handle extreme values.

> For our purposes, it is always enabled.

Author: https://github.com/chhoumann/thesis-chemcam/pull/157#discussion_r1615779976

```suggestion
```

> Moreover, we determined that utilizing a percentile value of $p=5\%$ strikes an optimal balance between maintaining the representativeness of the test dataset and ensuring that all highly influential targets are included in the training dataset.

Reviewer: I agree with Patrick on this. A least be clear about the fact that this a somewhat rudimentary approach.

> Moreover, we determined that utilizing a percentile value of $p=5\%$ strikes an optimal balance between maintaining the representativeness of the test dataset and ensuring that all highly influential targets are included in the training dataset.

Author: ![image](https://github.com/chhoumann/thesis-chemcam/assets/29108628/f6e1cb30-d6c1-4612-b1bf-6dfea0e1b91b)

> Moreover, we determined that utilizing a percentile value of $p=5\%$ strikes an optimal balance between maintaining the representativeness of the test dataset and ensuring that all highly influential targets are included in the training dataset.

Author: Obviously I'm kidding. That's why I implemented the slider for the streamlit page. Wanted to check & validate that this was actually correct. I checked the composition values for various configurations and found that exactly 5% was the optimal value.

> Moreover, we determined that utilizing a percentile value of $p=5\%$ strikes an optimal balance between maintaining the representativeness of the test dataset and ensuring that all highly influential targets are included in the training dataset.

Author: Let's move all our specific configurations to the methodology section as discussed.

```suggestion
```

The fold assignments are then merged back into the original dataset (Line~\ref{line:merge_folds}), and the dataset is split into a test set (consisting of the highest fold number) and the remaining training data (Line~\ref{line:split_dataset}).
The training data is further divided into \( k-1 \) folds for cross-validation.
For each fold, the training data consists of all but one fold, which serves as the validation set.
These pairs of training and validation sets are then appended to the list of cross-validation folds \(\mathbf{F}_\text{cv}\) (Line~\ref{line:create_folds}).

Finally, the fold indicator column is removed from all datasets before returning the final partitions (Line~\ref{line:remove_fold_column}).
This cleanup step ensures that the fold information does not interfere with subsequent data processing or model training.

The final output of this procedure consists of:
\begin{itemize}
    \item The cross-validation folds \(\mathbf{F}_\text{cv}\), each containing a tuple of training and validation sets.
    \item The training set \(\mathbf{D}_\text{train}\) in its entirety.
    \item The test set \(\mathbf{D}_\text{test}\), distinct from the training set.
\end{itemize}

> 

Reviewer: I agree with Patrick that this could be pretty neat.

> 

Author: What if I add this?

![image](https://github.com/chhoumann/thesis-chemcam/assets/29108628/eaf23e8c-9744-42dc-b25c-c8cfc9966873)

And then also include the figure showing the distributions before and after as well? This one:
![original_and_post_fold](https://github.com/chhoumann/thesis-chemcam/assets/29108628/bd2723fe-244f-4787-a452-1c20d5fbff33)

> 

Author: Adding after discussion

Our method is broadly aligned with the approach described by \citet{andersonImprovedAccuracyQuantitative2017}.
They employed a similar strategy to assess the performance of their PLS model, using k-fold cross-validation and a separate test set.
Their process involved dividing the full set of laboratory data into five folds, using four for cross-validation and combining them as the final training set, while the fifth fold served as a test set.
This approach ensured that each fold represented the full elemental compositional variation, and extreme targets were included in the training folds to handle a wider range of compositions.
For consistency, we also use \(k=5\) for our cross-validation.

Additionally, by using $k=5$ folds, we have effectively chosen an 80\%/20\% split between the training and testing datasets.
In our experience, this ratio maximizes the training set's capacity for effective model learning while ensuring that the testing set is sufficiently representative to provide an accurate assessment of the model's performance on new data.

> As detailed in Section~\ref{sec:problem_definition}, the limited availability of data is a primary constraint.

Author: Done

> Allocating too much data to the testing set could compromise the comprehensiveness of the training set, undermining the model's ability to generalize effectively.

Author: That was actually being mentioned in the Evaluation Metrics section. We can add the following line to further clarify the implications.

But just to clarify, we actually get higher RMSE_CV values, as the 'hard to predict values' are placed in the training set, which the folds based on. The RMSEP (RMSE on test set) is actually lower, since it doesn't deviate much from the mean.

```diff
However, our data partitioning method, which moves the most extreme values into the training data, naturally results in the testing data being closer to the mean of the data distribution, making it easier to predict.
+ In practice, this would result in lower \texttt{rmsep} and \texttt{std\_dev} values compared to the cross validation metrics.
Therefore, evaluating the model's performance using both the cross-validation metrics (\texttt{rmse\_cv} and \texttt{std\_dev\_cv}) and the test set metrics (\texttt{rmsep} and \texttt{std\_dev}) is crucial.
The cross-validation metrics provide insights into the model's stability across different subsets, while the test set metrics offer a final measure of performance on truly unseen data, giving a comprehensive assessment of the model's generalizability.
```

> Allocating too much data to the testing set could compromise the comprehensiveness of the training set, undermining the model's ability to generalize effectively.

Author: Added

    \caption{Combined Histogram and \gls{kde} of \ce{SiO_2} Distribution in Each Fold.}
    \label{fig:histogram_kde_plot}
\end{figure*}

Figure \ref{fig:histogram_kde_plot} combines the histograms and \gls{kde} curves of \ce{SiO_2} values from each fold and the full test set into a single plot, using different colors to represent each fold and the test set.
This visualization enables direct comparison of the distributions across the folds and the test set.

The combined histograms and \gls{kde} curves show a high degree of overlap, reinforcing the observation that the distributions are consistent across different folds and the test set.
The test set distribution closely follows the overall pattern observed in the folds, indicating that the test set is also representative of the overall dataset.
The presence of extreme values is evident in the tails of the distributions, particularly in the training folds.

\subsubsection{Statistical Consistency}

To further validate our visual analysis, we can look at quantitative measures such as the means and standard deviations of \ce{SiO_2} concentrations across the folds and the overall dataset.

> As shown in Table~\ref{tab:siO2_std_means}, the standard deviations and means for \ce{SiO_2} concentrations in each fold are as follows:

Author: ![image](https://github.com/chhoumann/thesis-chemcam/assets/29108628/7cea5aa8-68f6-442e-a777-87dac8720cea)

Does for me now. We can fix later if additional text in the report messes the formatting.

\begin{table}[h!]
    \centering
    \begin{tabular}{|c|c|c|}
        \hline
        \textbf{Fold} & \textbf{Standard Deviation} & \textbf{Mean} \\
        \hline
        Fold 1 & 16.28 & 55.70 \\
        \hline
        Fold 2 & 12.46 & 57.59 \\
        \hline
        Fold 3 & 18.01 & 51.51 \\
        \hline
        Fold 4 & 15.77 & 55.10 \\
        \hline
        Test Full & 9.06 & 56.57 \\
        \hline
    \end{tabular}
    \caption{Standard deviations and means of \ce{SiO_2} concentrations across different folds' validation sets and the full test set.}
    \label{tab:siO2_std_means}
\end{table}

> The means and standard deviations of \ce{SiO_2} concentrations for each fold are consistent with those of the overall dataset.

Author: Added in 98b16ada5c11

These metrics are calculated for each fold and averaged across all folds to provide comprehensive indicators of model accuracy and variability.
In addition, we also compute the metrics for the test set to provide a measure of the model's performance on unseen data.
Therefore, we have the following metrics for each experiment:
\begin{enumerate}
    \item \textbf{Fold-specific \gls{rmse} and Standard Deviation:} For each of the $k$ folds, we calculate both the \gls{rmse} and standard deviation, denoted as \texttt{rmse\_cv\_n} and \texttt{std\_dev\_cv\_n}, where \texttt{n} ranges from 1 to $k$.
    \item \textbf{Average \gls{rmse} and Standard Deviation:} The overall cross-validation \gls{rmse} (\texttt{rmse\_cv}) and standard deviation (\texttt{std\_dev\_cv}) are computed as the mean of the fold-specific values.
    \item \textbf{Test Set \gls{rmse} and Standard Deviation:} The \gls{rmse} and standard deviation are also computed for the test set, denoted as \texttt{rmsep} and \texttt{std\_dev}, to provide a measure of the model's performance on unseen data.

> \end{enumerate}

Author: Nice, thanks!

Therefore, evaluating the model's performance using both the cross-validation metrics (\texttt{rmse\_cv} and \texttt{std\_dev\_cv}) and the test set metrics (\texttt{rmsep} and \texttt{std\_dev}) is crucial.
The cross-validation metrics provide insights into the model's stability across different subsets, while the test set metrics offer a final measure of performance on truly unseen data, giving a comprehensive assessment of the model's generalizability.

\subsubsection{Conclusion}
In summary, our validation and testing procedures are designed to ensure the robustness and generalizability of our models by employing a combination of k-fold cross-validation and a separate test set.
The customized k-fold cross-validation algorithm mitigates data leakage by ensuring that data points from the same target are not split across training and testing sets.
This method, which includes handling extreme values, aligns with established practices and optimally partitions the dataset to maintain a balance between representativeness and model training efficacy.

> By assessing model performance through comprehensive metrics such as \gls{rmse} and standard deviation across both cross-validation folds and a separate test set, we provide a rigorous evaluation framework that underscores the reliability and accuracy of our predictive models.

Reviewer 2: ```suggestion
By assessing model performance through metrics such as \gls{rmse} and standard deviation across both cross-validation folds and a separate test set, we provide an evaluation framework that underscores the reliability and accuracy of our predictive models.
```

Maybe tone it down a little bit with the adjectives haha

Author: 😂😂

For consistency, we also use \(k=5\) for our cross-validation.

Additionally, by using $k=5$ folds, we have effectively chosen an 80\%/20\% split between the training and testing datasets.
In our experience, this ratio maximizes the training set's capacity for effective model learning while ensuring that the testing set is sufficiently representative to provide an accurate assessment of the model's performance on new data.
Allocating too much data to the testing set could compromise the comprehensiveness of the training set, undermining the model's ability to generalize effectively due to the limited availability of data.

\subsubsection{Visual Analysis of \ce{SiO_2} Distribution in Folds}
This section aims to provide a detailed visual and statistical analysis of the \ce{SiO_2} concentration distribution across different folds used in our custom k-fold data partitioning procedure.
The purpose is to demonstrate the effectiveness of our strategy in maintaining balanced and representative datasets, which is crucial for developing robust predictive models.

We focus specifically on \ce{SiO_2} as a representative example to illustrate our methodology.

> Similar analyses were conducted for other oxides; however, the plots and figures for these are omitted here for brevity.

Author: That could be. Let's ask them on Thursday. Adding as point on agenda.

> \begin{figure*}[h!]

Author: As mentioned in the other comment: not much to do about it at this time.

    \label{tab:siO2_std_means}
\end{table}

The means and standard deviations of \ce{SiO_2} concentrations for each fold are consistent with those of the overall dataset.
These metrics are even more similar across folds for the training sets, indicating that the training sets are also representative of the entire dataset.
This quantitative consistency supports the visual evidence that each fold is a reliable representative of the entire dataset.

\subsubsection{Importance of Distribution Balance}
Maintaining balanced and representative distributions in each fold is essential for training robust models.
It ensures that the models are not biased towards any specific subset of the data, which is crucial for their generalization to unseen data.
The balanced distributions across folds enable the model to be trained on diverse subsets of data, leading to better generalization.
This approach reduces the risk of overfitting and ensures that the model performs well on new, unseen data.

By ensuring that each fold is representative of the overall dataset, these visualizations validate our dataset partitioning strategy.

> The consistency observed across folds supports the claim that the data distribution method works as intended, providing confidence in the generalizability and reliability of our models.

Reviewer 2: ```suggestion
The consistency observed across folds further supports the claim that the data distribution method works as intended, providing confidence in the generalizability and reliability of our models.
```
If you decide to apply the above suggestion.

\begin{table}[h!]
    \centering
    \begin{tabular}{|c|c|c|}
        \hline
        \textbf{Fold} & \textbf{Standard Deviation} & \textbf{Mean} \\
        \hline
        Fold 1 & 16.28 & 55.70 \\
        \hline
        Fold 2 & 12.46 & 57.59 \\
        \hline
        Fold 3 & 18.01 & 51.51 \\
        \hline
        Fold 4 & 15.77 & 55.10 \\
        \hline
        Train Full & 15.91 & 54.96 \\
        \hline
        Test Full & 9.06 & 56.57 \\
        \hline
        Full & 14.94 & 55.25 \\
        \hline
    \end{tabular}
    \caption{Standard deviations and means of \ce{SiO_2} concentrations across different folds' validation sets, the full training set, the full test set, and the entire dataset.}
    \label{tab:siO2_std_means}
\end{table}

From the table, it is evident that the means and standard deviations of \ce{SiO_2} concentrations for each fold are consistent with those of the overall dataset.
These metrics are even more similar across folds for the training sets, indicating that the training sets are also representative of the entire dataset.
This quantitative consistency supports the visual evidence that each fold is a reliable representative of the entire dataset.

> \subsubsection{Importance of Distribution Balance}

Reviewer: ```suggestion

```
